examples/window-app/common/include/WindowApp.h

```diff
@@ -0,0 +1,145 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/common/gen/enums.h>
+#include <core/CHIPError.h>
+
+class WindowApp
+{
+public:
+    struct Timer
+    {
+        typedef void (*Callback)(Timer & timer);
+
+        Timer(const char * name, uint32_t timeoutInMs, Callback callback, void * context) :
+            mName(name), mTimeoutInMs(timeoutInMs), mCallback(callback), mContext(context)
+        {}
+        virtual ~Timer()     = default;
+        virtual void Start() = 0;
+        virtual void Stop()  = 0;
```

Comment:
What are the rules for using virtual functions? I don't think that these are suitable for small embedded devices?

Reply:
Virtual functions are allowed in CHIP as long as they make sense. There is no rule of "everything virtual because testing" (that would be too much for sure) however we do not prevent virtual either. Overhead is assumed very small and may be even better than manually trying to create structs containing function pointers (compiler vtable is better than a manual one)